dali/python/nvidia/dali/ops/_signatures.py

```diff
@@ -131,15 +131,92 @@ def _get_positional_input_param(schema, idx):
                          annotation=annotation)
 
 
-def _get_positional_input_params(schema):
+def _get_annotation_input_regular(schema):
+    """Return the annotation for regular input parameter in DALI, used for the primary overload.
+    A function is used as a global variable can be confused with type alias.
+    TODO(klecki): Extend with TensorLike.
+    """
+    return _DataNode
+
+
+def _get_annotation_return_regular(schema):
+    """Produce the return annotation for DALI operator suitable for primary, non-MIS overload.
+    Note the flattening, single output is not packed in Sequence.
+    """
+    if schema.HasOutputFn():
+        # Dynamic number of outputs, not known at "compile time"
+        return_annotation = Union[_DataNode, Sequence[_DataNode], None]
+    else:
+        # Call it with a dummy spec, as we don't have Output function
+        num_regular_output = schema.CalculateOutputs(_b.OpSpec(""))
+        if num_regular_output == 0:
+            return_annotation = None
+        elif num_regular_output == 1:
+            return_annotation = _DataNode
+        else:
+            # Here we could utilize the fact, that the tuple has known length, but we can't
+            # as DALI operators return a list
+            # Also, we don't advertise the actual List type, hence the Sequence.
+            return_annotation = Sequence[_DataNode]
+    return return_annotation
+
+
+def _get_annotation_input_mis(schema):
+    """Return the annotation for multiple input sets, used for the secondary operator overload.
+    A function is used as a global variable can be confused with type alias.
+    Handles special case for operators with one input.
+    """
+    if schema.MinNumInput() == 1 and schema.MaxNumInput() == 1:
+        return List[_DataNode]
+    return Union[_DataNode, List[_DataNode]]
```

Comment:
I am confused a little here (likely not seeing the context) - we do we need union with `_DataNode` when the num inputs is not exactly one, but can get rid of it otherwise?

Reply:
So, we have primary overload of the form:
```
def operator(input0: DataNode, input1: DataNode, ...) -> ...: ...
```
and the secondary one:
```
def operator(input0: Union[DataNode, List[DataNode]], input1: Union[DataNode, List[DataNode]], ...) -> ...: ...
```
The first catches the case where all inputs are DataNodes, the second will catch cases where at least one of the input is a list (the rest of the inputs can still be DataNodes and can be broadcast to match the length of the list).

If we have just one input, we have two overloads:
```
def operator(input0: DataNode, /, ...) -> ...: ...
def operator(input0: List[DataNode], /, ...) -> ...: ...
```
we don't heed the union in the second one, as the only DataNode inputs always matches the first overload.

Reply:
Oh, I see, I was missing the broadcasting case in my head. Thank you!